task/control-plane-agent/src/dump.rs

```diff
@@ -0,0 +1,221 @@
+                .map_err(|_e| SpError::Dump(DumpError::BadArea))?;
+            header = humpty::DumpAreaHeader::read_from(
+                &data[..size_of::<humpty::DumpAreaHeader>()],
+            )
+            .unwrap_lite();
+            if header.contents == humpty::DumpContents::SingleTask.into()
+                && header.nsegments > 0
+            {
+                if count == dump_index {
+                    found = Some(index);
+                    break;
+                } else {
+                    count += 1;
+                }
+            }
+            if header.next == 0 {
+                break;
+            }
+        }
+        let Some(index) = found else {
+            return Err(SpError::Dump(DumpError::BadIndex));
+        };
+
+        // Here we go!
+        let offset = size_of::<humpty::DumpAreaHeader>()
+            + usize::from(header.nsegments)
+                * size_of::<humpty::DumpSegmentHeader>();
+        let task = humpty::DumpTask::read_from_prefix(&data[offset..])
+            .ok_or(SpError::Dump(DumpError::NoDumpTaskHeader))?;
+        if task.magic != humpty::DUMP_TASK_MAGIC {
+            return Err(SpError::Dump(DumpError::CorruptTaskHeader));
+        }
+
+        // Pick a client slot to use.  Prefer empty slots or slots that used the
+        // same key (which is bad behavior on the client's part); otherwise,
+        // just use slot 0.
+        let slot = self
+            .clients
+            .iter()
+            .position(|c| c.map(|c| c.key == key).unwrap_or(true))
+            .unwrap_or(0);
+        self.clients[slot] = Some(ClientDumpState {
+            key,
+            area_index: index,
+            offset: (offset + size_of::<humpty::DumpTask>()) as u32,
+        });
+
+        Ok(DumpTask {
+            task: task.id,
+            time: task.time,
+        })
+    }
+
+    fn clear_client_state(&mut self, key: u32) {
+        self.clients
+            .iter_mut()
+            .filter(|c| c.map(|c| c.key == key).unwrap_or(false))
+            .for_each(|c| *c = None)
+    }
+
+    pub(crate) fn task_dump_read_continue(
+        &mut self,
+        key: u32,
+        buf: &mut [u8],
+    ) -> Result<Option<DumpSegment>, SpError> {
+        let Some(state) =
+            self.clients.iter_mut().flatten().find(|c| c.key == key)
+        else {
+            return Err(SpError::Dump(DumpError::BadKey));
+        };
+
+        let mut header = humpty::DumpAreaHeader::new_zeroed();
+        self.agent
+            .read_dump_into(state.area_index, 0, header.as_bytes_mut())
+            .map_err(|_e| SpError::Dump(DumpError::ReadFailed))?;
+
+        // Make sure the header is still valid
+        if header.contents != humpty::DumpContents::SingleTask.into() {
+            return Err(SpError::Dump(DumpError::NoLongerValid));
+        }
+
+        // Move along to the next area if we're at the end
+        if state.offset + size_of::<humpty::DumpSegmentData>() as u32
```

Comment:
considered obnoxious, feel free to ignore: `size_of::<humpty::DumpSegmentData>()` and `size_of::<humpty::DumpAreaHeader>()` are a little bit hard to tell apart in this function, because they're both long-ish expressions and the type parameter to `size_of` starts with `humpty::Dump` and my eyes glazed over at that point. I _might_ consider doing something like
```rust
const HEADER_SIZE: u32 = size_of::<humpy::DumpAreaHeader>() as u32;
const SEGMENT_DATA_SIZE: u32 = size_of::<humpy::DumpSegmentData>() as u32;
```
and then referring to them by those constants elsewhere, so that it's more obvious that they're different things? but maybe this is just a "me thing", in which case you're welcome to ignore me.